lib/engine/hex.rb

```diff
@@ -110,20 +110,38 @@ def lay(tile)
           @tile.cities.zip(tile.cities).to_h
         # if @tile is not blank, ensure connectivity is maintained
         else
-          @tile.cities.map.with_index do |old_city, index|
+          @tile.cities.map do |old_city|
             new_city = tile.cities.find do |city|
               # we want old_edges to be subset of new_edges
               # without the any? check, first city will always match
-              old_city.exits.any? && (old_city.exits - city.exits).empty?
+              !old_city.exits.empty? && (old_city.exits - city.exits).empty?
             end
 
-            # When downgrading from yellow to no-exit tiles, assume it's the same index
-            # Also, when upgrading a no-exit city, assume it's the same index if possible
-            new_city ||= (tile.cities[index] || tile.cities[0])
             [old_city, new_city]
           end.to_h
         end
 
+      # When downgrading from yellow to no-exit tiles, assume it's the same index
+      # Also, when upgrading a no-exit city, assume it's the same index if possible, otherwise
+      # pick first available city
+      new_cities = city_map.values.compact
+      @tile.cities.each_with_index do |old_city, index|
+        next if city_map[old_city]
+
+        new_city = if !new_cities.include?(tile.cities[index])
+                     tile.cities[index]
+                   elsif !new_cities.include?(tile.cities[0])
+                     tile.cities[0]
+                   else
+                     tile.cities.find { |city| !new_cities.include?(city) }
+                   end
```

Comment:
the code smells to me, why would the fail safe ever hit?
new_city = tile.cities[index]
new_city = tile.cities.find { |city| !new_cities.include?(city) if new_cities.include?(new_city)

Reply:
I think the only time it could hit is when we map from N cities on the old tile to less then N cities on the new tile.

Since I don't know the reason why the original code had
new_city ||= (tile.cities[index] || tile.cities[0])
I was hesitant to change it.